src/mame/konami/konppc.h

```diff
@@ -38,70 +43,70 @@ class konppc_device :  public device_t
-
-	void nwk_fifo_0_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void nwk_fifo_1_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	uint32_t nwk_voodoo_0_r(offs_t offset);
-	uint32_t nwk_voodoo_1_r(offs_t offset);
-	void nwk_voodoo_0_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
-	void nwk_voodoo_1_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0);
+	template <unsigned Board> uint32_t cgboard_comm_sharc_r(offs_t offset) { return dsp_comm_sharc_r(Board, offset); }
+	template <unsigned Board> void cgboard_comm_sharc_w(offs_t offset, uint32_t data) { dsp_comm_sharc_w(Board, offset, data); }
+	template <unsigned Board> uint32_t cgboard_shared_sharc_r(offs_t offset) { return dsp_shared_ram_r_sharc(Board, offset); }
+	template <unsigned Board> void cgboard_shared_sharc_w(offs_t offset, uint32_t data) { dsp_shared_ram_w_sharc(Board, offset, data); }
+
+	template <unsigned Board> uint32_t cgboard_k033906_r(offs_t offset) { return k033906_r(Board, offset); }
+	template <unsigned Board> void cgboard_k033906_w(offs_t offset, uint32_t data) { k033906_w(Board, offset, data); }
+
+	template <unsigned Board> void nwk_voodoo_fifo_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0) { voodoo_fifo_w(Board, offset, data, mem_mask); }
+	template <unsigned Board> uint32_t nwk_voodoo_r(offs_t offset) { return voodoo_r(Board, offset); }
+	template <unsigned Board> void nwk_voodoo_w(offs_t offset, uint32_t data, uint32_t mem_mask = ~0) { voodoo_w(Board, offset, data, mem_mask); }
+
 protected:
 	// device-level overrides
 	virtual void device_start() override ATTR_COLD;
 
-	uint32_t dsp_comm_sharc_r(int board, int offset);
-	void dsp_comm_sharc_w(int board, int offset, uint32_t data);
-	uint32_t dsp_shared_ram_r_sharc(int board, int offset);
-	void dsp_shared_ram_w_sharc(int board, int offset, uint32_t data);
-
-	uint32_t nwk_fifo_r(int board);
-	void nwk_fifo_w(int board, uint32_t data);
 private:
 	// device finders
 	optional_device_array<adsp21062_device, 2> m_dsp;
 	optional_device_array<k033906_device, 2> m_k033906;
 	optional_device_array<generic_voodoo_device, 2> m_voodoo;
 
 	// internal state
-	uint32_t dsp_comm_ppc[MAX_CG_BOARDS][2];
-	uint32_t dsp_comm_sharc[MAX_CG_BOARDS][2];
-	uint8_t dsp_shared_ram_bank[MAX_CG_BOARDS];
+	uint32_t m_dsp_comm_ppc[MAX_CG_BOARDS][2];
+	uint32_t m_dsp_comm_sharc[MAX_CG_BOARDS][2];
+	uint8_t m_dsp_shared_ram_bank[MAX_CG_BOARDS];
 
-	int32_t cgboard_id = 0;
-	int32_t cgboard_type;
-	int32_t num_cgboards;
+	int32_t m_cgboard_id = 0;
+	int32_t m_cgboard_type;
+	int32_t m_num_cgboards;
 
-	std::unique_ptr<uint32_t[]> dsp_shared_ram[MAX_CG_BOARDS];
+	std::unique_ptr<uint32_t[]> m_dsp_shared_ram[MAX_CG_BOARDS];
 
-	uint32_t dsp_state[MAX_CG_BOARDS]{};
-	uint32_t nwk_device_sel[MAX_CG_BOARDS]{};
-	const char *texture_bank[MAX_CG_BOARDS]{};
+	uint32_t m_dsp_state[MAX_CG_BOARDS]{};
+	uint32_t m_nwk_device_sel[MAX_CG_BOARDS]{};
+	memory_bank *m_texture_bank[MAX_CG_BOARDS]{};
```

Comment:
Would it be a good idea to use `optional_memory_bank_array<2>` for `m_texture_bank` and allow the system using it to configure the tags?

Reply:
but nwk-tr.cpp uses both SHARC banks are mapped as board 0, it will be not as easy...
```
void nwktr_state::init_racingj()
{
	init_nwktr();
	m_konppc->set_cgboard_texture_bank(0, membank("master_cgboard_bank"), memregion("master_cgboard")->base());
	m_konppc->set_cgboard_texture_bank(0, membank("slave_cgboard_bank"), memregion("slave_cgboard")->base()); // for some reason, additional CG roms are located on the slave CG board...
}

void nwktr_state::init_thrilld()
{
	init_nwktr();
	m_konppc->set_cgboard_texture_bank(0, membank("master_cgboard_bank"), memregion("master_cgboard")->base());
	m_konppc->set_cgboard_texture_bank(0, membank("slave_cgboard_bank"), memregion("master_cgboard")->base()); // ...while this is not the case for thrilld
}
```

Reply:
It's a copy/paste error on my end. The slave bank should have been set to 1 and I failed to notice that small detail.